File: pluralsight/applyingfunctionalprincipalsincsharp/avoidingnullswiththemaybetype/AvoidingNulls.Core/Result.cs
namespace FunctionalPrinciples.AvoidingNulls.Core
{
    public class Result
    {
        public bool IsSuccess { get; }

        public bool IsFailure => !string.IsNullOrEmpty(Error);

        public string Error { get; }

        protected Result(bool isSuccess, string error)
        {
            IsSuccess = isSuccess;
            Error = error;
        }

        public static Result Ok()
        {
            return new Result(true, string.Empty);
        }

        public static Result<T> Ok<T>(T value)
        {
            return new Result<T>(value, true, string.Empty);
        }

        public static Result Fail(string error)
        {
            return new Result(false, error);
        }

        public static Result<T> Fail<T>(string error)
        {
            return new Result<T>(default(T), false, error);
        }
    }

    public class Result<T> : Result
    {
        public T Value { get; }

        public Result(T value, bool isSuccess, string error) : base(isSuccess, error)
        {
            Value = value;
        }
    }
}

File: alura/course_android_part_001/app/src/main/java/br/com/flaviogf/schedule/infrastructure/Result.java
package br.com.flaviogf.schedule.infrastructure;

public class Result<T> {
    private final Boolean isSuccess;
    private final String message;

    private Result(boolean isSuccess, String message) {
        this.isSuccess = isSuccess;
        this.message = message;
    }

    public boolean isSuccess() {
        return isSuccess;
    }

    public boolean isFailure() {
        return !isSuccess;
    }

    public String getMessage() {
        return message;
    }

    public static <T> Result<T> ok() {
        return new Result<>(true, "");
    }

    public static <T> Result<T> fail(String message) {
        return new Result<>(false, message);
    }
}
